### PR TITLE
Test/firestore more tests

### PR DESCRIPTION
**Pull request by userdarius**
# Firebase Testing Implementation

## Description
This PR implements comprehensive unit tests for Firebase Firestore operations, covering CRUD operations for various entities in our application. The tests ensure reliable data handling and validation across our Firebase infrastructure.

## Key Implementations
- Added unit tests for core entities:
  - Users (update)
  - Residences (create, get, update, delete)
  - Apartments (create, get, update, delete)
  - Maintenance Requests (create, get, update, delete)
  - Laundry Machines (create, get, update, delete)
- Implemented validation testing for tenant codes
- Added tests for utility functions like `generate_unique_code`, `validateTenantCode`, and `deleteUsedTenantCodes`
- Comprehensive error handling tests for invalid inputs and failed operations

## Notable Exclusions and fixes
- Bug fix introduced during last PR merge to main of TUser -> User type declarations in certain files causing errors.
- Tests for `add_new_landlord` and `add_new_tenant` functions are intentionally omitted as they are being handled in a separate PR.

## Testing Approach
- Used Jest mocking to simulate Firestore operations
- Implemented proper cleanup after each test
- Covered both success and failure scenarios
- Included validation for required fields and data integrity

## Impact
This PR significantly improves our test coverage for Firebase operations, ensuring more reliable data handling and reducing the risk of runtime errors in production.

### CODE REVIEW

File: app/screens/auth/SignUpScreen.tsx
Comment (baudoin-csp) [inline app/screens/auth/SignUpScreen.tsx:1]: Thanks for fixing this typo!

File: firebase/firestore/firestore.ts
Comment (baudoin-csp) [inline firebase/firestore/firestore.ts:1]: It seems like you found some bugs and edge cases! Always a good idea to check if IDs are empty strings or null.

File: app/__tests__/unit/firebase.test.tsx
Comment (baudoin-csp) [inline app/__tests__/unit/firebase.test.tsx:1]: Maybe we can think about importing all functions at once from ```firebase/firestore``` using * import

Comment (baudoin-csp) [reply app/__tests__/unit/firebase.test.tsx:1]: Thanks for using only one mock for each scenario, it simplifies the review of the code and is more efficient!

Comment (baudoin-csp) [reply app/__tests__/unit/firebase.test.tsx:1]: The updates to the unit tests for Firebase functions in this file are comprehensive and well-structured. The use of mocks allows for isolated testing of each function. Adding tests for error scenarios increases the robustness of the test suite. Overall, these changes improve the reliability and maintainability of the codebase by ensuring critical functions are thoroughly tested.

Comment (baudoin-csp) [reply app/__tests__/unit/firebase.test.tsx:1]: ### PR Review

The PR looks solid overall, you’ve got a good mix of positive and negative test cases. One thing that stood out was how well you handled edge cases like empty or invalid inputs, especially in functions like `getUser` and `deleteUser`. That’s something a lot of people tend to overlook, so kudos for covering those scenarios. However, in some places, like `generate_unique_code`, you might want to test for uniqueness more rigorously. I mean, it's called “unique” code, but there’s no test to prove that two calls don’t generate the same code— that's not a big deal right but might be worth adding in the future.

Another thing, the mocks are well-structured, is it possible to centralize it more? And I did notice a couple of spots where `toHaveBeenCalledTimes` could add more precision. Not a big deal, but it’d make your tests a bit more bulletproof.

**Review by baudoin-csp — APPROVED** [review]
The PR looks solid overall, you’ve got a good mix of positive and negative test cases. One thing that stood out was how well you handled edge cases like empty or invalid inputs, especially in functions like `getUser` and `deleteUser`. That’s something a lot of people tend to overlook, so kudos for covering those scenarios. However, in some places, like `generate_unique_code`, you might want to test for uniqueness more rigorously. I mean, it's called “unique” code, but there’s no test to prove that two calls don’t generate the same code— that's not a big deal right but might be worth adding in the future.

Another thing, the mocks are well-structured, is it possible to centralize it more? And I did notice a couple of spots where `toHaveBeenCalledTimes` could add more precision. Not a big deal, but it’d make your tests a bit more bulletproof.

**Comment by userdarius** [comment]
Will keep these comments into consideration. Thanks a lot for the review!

**Pull request merged**